node/sync/src/ping.rs

```diff
@@ -0,0 +1,198 @@
+    const MAX_PING_INTERVAL: Duration = Duration::from_secs(20);
+
+    /// Create a new instance of the ping logic.
+    /// There should only be one per node.
+    ///
+    /// # Usage
+    /// Initialize this with the most up-to-date block locators and call
+    /// update_block_locators, whenever a new block is received/created.
+    pub fn new(router: Router<N>, block_locators: BlockLocators<N>) -> Self {
+        let notify = Arc::new(Notify::default());
+        let inner = Arc::new(Mutex::new(PingInner::new(Some(block_locators))));
+
+        {
+            let inner = inner.clone();
+            let router = router.clone();
+            let notify = notify.clone();
+
+            tokio::spawn(async move {
+                Self::ping_task(&inner, &router, &notify).await;
+            });
+        }
+
+        Self { inner, router, notify }
+    }
+
+    /// Same as [`Self::new`] but for nodes that peers cannot sync from
+    /// such as provers.
+    pub fn new_nosync(router: Router<N>) -> Self {
+        let notify = Arc::new(Notify::default());
+        let inner = Arc::new(Mutex::new(PingInner::new(None)));
+
+        {
+            let inner = inner.clone();
+            let router = router.clone();
+            let notify = notify.clone();
+
+            tokio::spawn(async move {
+                Self::ping_task(&inner, &router, &notify).await;
+            });
+        }
+
+        Self { inner, router, notify }
+    }
+
+    /// Notify the ping logic that we received a Pong response.
+    pub fn on_pong_received(&self, peer_ip: SocketAddr) {
+        let now = Instant::now();
+        let mut inner = self.inner.lock();
+
+        inner.next_ping.insert(now + Self::MAX_PING_INTERVAL, peer_ip);
+
+        // self.notify.notify() is not needed as ping_task wakes up every MAX_PING_INTERVAL
+    }
+
+    /// Notify the ping logic that a new peer connected.
+    pub fn on_peer_connected(&self, peer_ip: SocketAddr) {
+        // Send the first ping.
+        let locators = self.inner.lock().block_locators.clone();
+        self.router.send_ping(peer_ip, locators);
+    }
+
+    /// Notify the ping logic that new blocks were created or synced.
+    pub fn update_block_locators(&self, locators: BlockLocators<N>) {
+        self.inner.lock().block_locators = Some(locators);
+
+        // wake up the ping task
+        self.notify.notify_one();
+    }
+
+    /// Background task that periodically sends out new ping messages.
+    async fn ping_task(inner: &Mutex<PingInner<N>>, router: &Router<N>, notify: &Notify) {
+        let mut new_block = false;
+
+        loop {
+            // Do not hold the lock while waiting.
+            let sleep_time = {
+                let mut inner = inner.lock();
+                let now = Instant::now();
+
+                // Ping peers.
+                if new_block {
+                    Self::ping_all_peers(&mut inner, router);
```

Comment:
Should this attempt to reset any respective entry in `inner.next_ping`, i.e. set the max ping interval after pinging?